tensorstore/kvstore/s3/s3_request_builder.cc

```diff
@@ -0,0 +1,283 @@
+#include "absl/strings/escaping.h"
+#include "absl/strings/str_split.h"
+#include "absl/status/status.h"
+#include "tensorstore/kvstore/s3/s3_request_builder.h"
+#include "tensorstore/kvstore/s3/validate.h"
+#include "tensorstore/internal/uri_utils.h"
+#include "tensorstore/internal/path.h"
+#include "tensorstore/internal/digest/sha256.h"
+#include "tensorstore/util/result.h"
+#include "tensorstore/util/str_cat.h"
+
+using ::tensorstore::internal::AsciiSet;
+using ::tensorstore::internal::ParseGenericUri;
+using ::tensorstore::internal::ParsedGenericUri;
+using ::tensorstore::internal::SHA256Digester;
+using namespace ::tensorstore::internal_http;
+using ::tensorstore::internal_storage_s3::IsValidBucketName;
+
+#ifndef TENSORSTORE_INTERNAL_S3_LOG_AWS4
+#define TENSORSTORE_INTERNAL_S3_LOG_AWS4 0
+#endif
+
+namespace tensorstore {
+namespace internal_storage_s3 {
+
+
+namespace {
+
+// See description of function UriEncode at this URL
+// https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
+constexpr AsciiSet kUriUnreservedChars{
+    "abcdefghijklmnopqrstuvwxyz"
+    "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+    "0123456789"
+    "-._~"};
+
+// NOTE: Only adds "/" to kUriUnreservedChars
+constexpr AsciiSet kUriKeyUnreservedChars{
+    "abcdefghijklmnopqrstuvwxyz"
+    "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+    "0123456789"
+    "/-._~"};
+
+
+/// Size of HMAC (size of SHA256 digest).
+constexpr static size_t kHmacSize = 32;
+
+void ComputeHmac(std::string_view key, std::string_view message, unsigned char (&hmac)[kHmacSize]) {
+    unsigned int md_len = kHmacSize;
+    // Computing HMAC should never fail.
+    ABSL_CHECK(HMAC(EVP_sha256(),
+                    reinterpret_cast<const unsigned char*>(key.data()),
+                    key.size(),
+                    reinterpret_cast<const unsigned char*>(message.data()),
+                    message.size(), hmac, &md_len) &&
+               md_len == kHmacSize);
+}
+
+void ComputeHmac(unsigned char (&key)[kHmacSize], std::string_view message, unsigned char (&hmac)[kHmacSize]){
+    unsigned int md_len = kHmacSize;
+    // Computing HMAC should never fail.
+    ABSL_CHECK(HMAC(EVP_sha256(), key, kHmacSize,
+                    reinterpret_cast<const unsigned char*>(message.data()),
+                    message.size(), hmac, &md_len) &&
+               md_len == kHmacSize);
+}
+
+
+} // namespace
+
+std::string UriEncode(std::string_view src) {
+  std::string dest;
+  internal::PercentEncodeReserved(src, dest, kUriUnreservedChars);
+  return dest;
+}
+
+std::string UriObjectKeyEncode(std::string_view src) {
+  std::string dest;
+  internal::PercentEncodeReserved(src, dest, kUriKeyUnreservedChars);
+  return dest;
+}
+
+S3RequestBuilder & S3RequestBuilder::AddHeader(std::string_view header) {
+  builder_.AddHeader(header);
+  return *this;
+}
+
+S3RequestBuilder & S3RequestBuilder::AddQueryParameter(std::string_view key, std::string_view value) {
+  builder_.AddQueryParameter(key, value);
```

Comment:
Here you could maintain the query parameters in your own vector. Below in BuildRequest you could do this:

```
std::stable_sort(std::begin(query_params_), std::end(query_params_));
for (const auto& [k, v] : query_params_) {
  builder_.AddQueryParameter(k, v);
}

auto request = builder_.BuildRequest();
```
Then your query parameters could just be extracted from the request url directly in CanonicalRequest().

Reply:
Thanks, this is great. Have used this approach.